appium/api-mocks/mock.ts

```diff
@@ -72,5 +73,51 @@ export class MockApi {
     }
   };
 
+  static get e2eMock() {
```

Comment:
Certain things will repeat a lot, and it will make sense to have a default for them. For example, the FES and EKM config.

Other things are specific for each test, and should be defined for each test. For example, google config and attester config.

I have very bad experience trying to define the state globally for all tests in one place. It gets confusing as to which test is using which parts of the config. Changing the config by adding one more message or one more contact may break another test, and so on.

Therefore we should do something like the following, for each test definition:

```ts
    const mockApi = new MockApi();
    mockApi.fesConfig = MockApiConfig.defaultEnterpriseFesConfiguration;
    mockApi.ekmConfig = MockApiConfig.defaultEnterpriseEkmConfiguration;
    mockApi.googleConfig = {
      accounts: {
        'e2e.enterprise.test@flowcrypt.com': {
          contacts: [CommonData.contacts.someParticularContact],
          messages: ['CC and BCC test'],
        }
      }
    };
    mockApi.attesterConfig = { 
      servedPubkeys: { 'expired@flowcrypt.com': attesterPublicKeySamples.expiredFlowcrypt }
    };
    mockApi.wkdConfig = {}
    await mockApi.withMockedApis(async () => {
      // test goes here
    });
```

Sure, it will be wordier. But it will be an absolute bliss to write and maintain (when you return to a particular test 4 years later) because you can look at the top of the test and know exactly how it is configured and why.

If many tests are using the same config, and the config contains messages only used by some of the tests, and contacts or public keys only used by some of the tests.. it becomes harder to dis-entangle it when understanding a particular test. And it also becomes less clear what to do when writing a new test - should I update the global configuration and risk another test will break? Or should I copy everything just to adjust one thing for a new test?

What I'm proposing is a system where there is just one way to do a particular thing: each test has its own minimal precise state definition, even if it means more lines of code overall.

Reply:
👍

Reply:
I'm ok to do it in a separate PR, if you'd like this one merged soon. Or it can be same PR - up to you.

Reply:
I'll do it in this PR, shouldn't take too much time.

Reply:
How about we implement this way?

interface MockApiOptions {
  fesConfig?: FesConfig;
  googleConfig?: GoogleConfig;
  wkdConfig?: WkdConfig;
  ekmConfig?: EkmConfig;
  attesterConfig?: FesConfig;
}
  constructor(options: MockApiOptions) {
    this.fesConfig = options.fesConfig || MockApiConfig.defaultEnterpriseFesConfiguration;
    this.googleConfig = options.googleConfig; // here to use default google config if needed like above like ( || MockApiConfig.defaultGoogleConfig)
    this.wkdConfig = options.wkdConfig || {};
    this.ekmConfig = options.ekmConfig || MockApiConfig.defaultEnterpriseEkmConfiguration;
    this.attesterConfig = options.attesterConfig || {};
  }

Reply:
How about we implement this way?
This way we would above duplicating multiple lines of code in many places.

```ts
mockApi.fesConfig = MockApiConfig.defaultEnterpriseFesConfiguration; // this one
    mockApi.ekmConfig = MockApiConfig.defaultEnterpriseEkmConfiguration;  // this one
    mockApi.googleConfig = {
      accounts: {
        'e2e.enterprise.test@flowcrypt.com': {
          contacts: [CommonData.contacts.someParticularContact],
          messages: ['CC and BCC test'],
        }
      }
    };
    mockApi.attesterConfig = { 
      servedPubkeys: { 'expired@flowcrypt.com': attesterPublicKeySamples.expiredFlowcrypt }
    };
    mockApi.wkdConfig = {}  // this one

```
```ts
interface MockApiOptions {
  fesConfig?: FesConfig;
  googleConfig?: GoogleConfig;
  wkdConfig?: WkdConfig;
  ekmConfig?: EkmConfig;
  attesterConfig?: FesConfig;
}
  constructor(options: MockApiOptions) {
    this.fesConfig = options.fesConfig || MockApiConfig.defaultEnterpriseFesConfiguration;
    this.googleConfig = options.googleConfig; // here to use default google config if needed like above like ( || MockApiConfig.defaultGoogleConfig)
    this.wkdConfig = options.wkdConfig || {};
    this.ekmConfig = options.ekmConfig || MockApiConfig.defaultEnterpriseEkmConfiguration;
    this.attesterConfig = options.attesterConfig || {};
  }

const mockApi = new MockApi({
  attesterConfig: {
    servedPubkeys: { 'expired@flowcrypt.com': attesterPublicKeySamples.expiredFlowcrypt }
  }
});

mockApi.fesConfig = something; // if you want to change defaultFesConfig by constructor you can change it below init
```

Reply:
But imagine having this kind of duplicated codes in tens of places.
Also I don't think it's good to change class variable (better to make variable private and have public function for getter and setter).
With above suggestion, we can pass all required parameters in constructor and won't have above issue either

Reply:
> But imagine having this kind of duplicated codes in tens of places.

yes, but tests are different than implementation code. In tests, more explicit is better, to understand what is happening. More code is usually worse, but not always.

Also, many tests will be non-enterprise tests without FES or EKM. Maybe not as many as enterorise tests, but still. I much prefer to have it explicit.

> Also I don't think it's good to change class variable directly [like mockApi.ekmConfig = xx, mockApi.xx = xx...](better to make variable private and have public function for getter and setter).

We could have setters. Like

```ts
    mockApi.setFesConfig(MockApiConfig.defaultEnterpriseFesConfiguration);
    mockApi.setEkmConfig(MockApiConfig.defaultEnterpriseEkmConfiguration);
    mockApi.setGoogleConfig({
      accounts: {
        'e2e.enterprise.test@flowcrypt.com': {
          contacts: [CommonData.contacts.someParticularContact],
          messages: ['CC and BCC test'],
        }
      }
    });
    mockApi.setAttesterConfig({ 
      servedPubkeys: { 'expired@flowcrypt.com': attesterPublicKeySamples.expiredFlowcrypt }
    });
```

I guess we could have a little more compact way to set up google account:

```ts
    mockApi.setFesConfig(MockApiConfig.defaultEnterpriseFesConfiguration);
    mockApi.setEkmConfig(MockApiConfig.defaultEnterpriseEkmConfiguration);
    mockApi.setGoogleConfigWithOneAccount('e2e.enterprise.test@flowcrypt.com', {
      messages: ['CC and BCC test'],
      contacts: [CommonData.contacts.someParticularContact],
    });
    mockApi.setAttesterConfig({ 
      servedPubkeys: { 'expired@flowcrypt.com': attesterPublicKeySamples.expiredFlowcrypt }
    });
```

That would save some lines while keeping it explicit.

Reply:
But I still prefer to pass variables in constructor though rather than passing it via `setter` function or setting public variable.

Reply:
Okay, then I'll make these config variables private and add custom setters.

> But I still prefer to pass variables in constructor though rather than passing it via `setter` function or setting public variable.

In some tests we need to change these configs in the middle of the test execution. That's why we need ability to set configs not only in constructor.

Reply:
Let's leave current implementation for now, as it works well for tests and code will be the same for initialization and for updating mock config.

Reply:
I also think setters work pretty well for the consistency, whether you are creating a new mock api or updating it, the code is the same. I don't think the individual setters increase burden.